src/main/java/com/google/sps/servlets/ChatServlet.java

```diff
@@ -11,66 +11,194 @@
+
+        final String out = getOutputString("chatPage", data);
+
+        response.getWriter().println(out);
+    }
+
+    /**
+     * Is called by a button on the chat form. doPost takes the message and its
+     * attributes (such as timestamp and user) and stores them into Datastore.
+     */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        final String messageText = (String) request.getParameter(MESSAGE_TEXT_PROPERTY);
+        String groupID = (String) request.getParameter(GROUP_ID_PARAM);
+        final long timestamp = System.currentTimeMillis();
+
+        // This will be a placeholder until we can coordinate how to pass
+        // groupID from page to page
+        if (groupID == null) {
+            groupID = "123";
+        }
+
+        if (messageText.length() == 0) {
+            response.sendRedirect("/chat");
+            return;
+        }
+
+        // Reads API Key and HTTP referer from file 
+        File apiKeyFile = new File(classLoader.getResource("keys.txt").getFile());
+        Scanner scanner = new Scanner(apiKeyFile);
+        final String apiKey = scanner.nextLine();
+        final String referer = scanner.nextLine();
+        scanner.close();
+
+        final String perspectiveURL = 
+            "https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze?key=" + apiKey;
+
+        final Double commentScore = getCommentScore(perspectiveURL, referer, messageText);
+
+        // If the comment is not toxic, then post it to Datastore
+        if (commentScore <= COMMENT_SCORE_THRESHOLD) {
+            DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+            Entity messageEntity = new Entity(MESSAGE_KIND + groupID);
+            messageEntity.setProperty(MESSAGE_TEXT_PROPERTY, messageText);
+            messageEntity.setProperty(TIMESTAMP_PROPERTY, timestamp);
+            datastore.put(messageEntity);
+
+            response.sendRedirect("/chat");
+        } else {
+            ImmutableMap<String, String> errorData = ImmutableMap.of("errorMessage", 
+                "Your message contains content that may be deemed offensive by others. " +
+                "Please revise your message and try again.");
+
+            final String out = getOutputString("error", errorData);
+
+            response.getWriter().println(out);
+        }
+    }
+
+    /**
+     * Returns the output string for the response. In other words,
+     * it sets up the soy template with the passed in data.
+     */
+    private String getOutputString(String templateName, ImmutableMap data) {
+        
+        SoyFileSet sfs = SoyFileSet
+            .builder()
+            .add(new File(classLoader.getResource("chat.soy").getFile()))
+            .build();
+        SoyTofu tofu = sfs.compileToTofu();
+
+        return tofu.newRenderer("templates.chat." + templateName).setData(data).render();
+    }
```

Comment:
Could you pull this out to a utils class called SoyRendererUtils? Then it can take the filename, templatename, and data and can be used across all of the different web pages.